### PR TITLE
Migrate extra DSpace-CRIS "Live Import" sources to DSpace (EPO & ADS)

### LINKED ISSUE #3359: Migrate extra DSpace-CRIS "Live Import" sources to DSpace core
DSpace-CRIS (built/maintained by 4Science) has a number of [Live Import sources](https://wiki.lyrasis.org/display/DSDOC7x/Live+Import+from+external+sources) which are not yet available in DSpace core.  
These may include:
* Scopus
* CrossRef
* CiNii
* NASA/ADS
* Web of Science
* VuFind
* OpenAIRE (by Author and Title)

If possible, we should work with the 4Science team to contribute these back to core code, so that they can be available to and maintained by the entire DSpace community.

This can also benefit upcoming features which also build on Live Import / External Sources, like https://github.com/DSpace/DSpace/issues/2834 and https://github.com/DSpace/DSpace/issues/2833

### PR DESCRIPTION
**Pull request by Micheleboychuk**
## References
Fixes https://github.com/DSpace/DSpace/issues/3359

To facilitate code review, the work was divided into 4 PRs.
This PR contains the following connectors:
- **EpoImportMetadataSourceServiceImpl**
- **ADSImportMetadataSourceServiceImpl**

## Instructions for Reviewers
All the connectors present in this PR are not active by default, in particular, to work it is necessary to configure the apiKey in  dspace.cfg
**submission.lookup.ads.key =
for ADS (https://ui.adsabs.harvard.edu/help/api/)**
 and
**po.consumerKey =
epo.consumerSecretKey =
for EPO (https://developers.epo.org/)**

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/importer/external/ads/ADSImportMetadataSourceServiceImpl.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/ads/ADSImportMetadataSourceServiceImpl.java:25]: You might see if this is replaceable by `org.json.JSONArray`, as it'd mean we can remove this dependency from the POM

File: dspace-api/src/main/java/org/dspace/importer/external/ads/ADSImportMetadataSourceServiceImpl.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/ads/ADSImportMetadataSourceServiceImpl.java:115]: Please add JavaDocs to all private classes in this service

File: dspace-api/src/main/java/org/dspace/importer/external/epo/service/EpoImportMetadataSourceServiceImpl.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/epo/service/EpoImportMetadataSourceServiceImpl.java:263]: Please add JavaDocs to all private classes though!

File: dspace-api/src/main/java/org/dspace/importer/external/metadatamapping/contributor/JsonPathMetadataProcessor.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/metadatamapping/contributor/JsonPathMetadataProcessor.java:12]: Please add JavaDocs

File: dspace-api/src/main/java/org/dspace/importer/external/metadatamapping/contributor/SimpleJsonPathMetadataContributor.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/metadatamapping/contributor/SimpleJsonPathMetadataContributor.java:20]: Please add JavaDocs

File: dspace-api/src/main/java/org/dspace/importer/external/metadatamapping/contributor/SimpleJsonPathMetadataContributor.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/metadatamapping/contributor/SimpleJsonPathMetadataContributor.java:15]: Again, are we sure we need this JSONArray?  Or can we just use `org.json.JSONArray`?

File: dspace-api/src/main/java/org/dspace/importer/external/metadatamapping/contributor/SimpleXpathDateFormatMetadataContributor.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/metadatamapping/contributor/SimpleXpathDateFormatMetadataContributor.java:31]: Please add JavaDocs

File: dspace-api/src/main/java/org/dspace/importer/external/scopus/service/LiveImportClient.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/scopus/service/LiveImportClient.java:17]: Again, this should have its package changed. It's not specific to Scopus

File: dspace-api/src/main/java/org/dspace/importer/external/scopus/service/LiveImportClientImpl.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/scopus/service/LiveImportClientImpl.java:41]: Same here, package name should be changed

File: dspace-api/src/main/java/org/dspace/importer/external/scopus/service/LiveImportClient.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/scopus/service/LiveImportClient.java:19]: Needs JavaDocs

File: dspace-api/src/main/java/org/dspace/importer/external/scopus/service/LiveImportClient.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/scopus/service/LiveImportClient.java:21]: Needs JavaDocs

File: dspace-api/src/main/java/org/dspace/importer/external/scopus/service/LiveImportClientImpl.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/scopus/service/LiveImportClientImpl.java:125]: Please add JavaDocs

File: dspace-api/src/main/java/org/dspace/importer/external/scopus/service/LiveImportClientImpl.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/scopus/service/LiveImportClientImpl.java:136]: Please add JavaDocs

File: dspace-api/src/main/java/org/dspace/importer/external/scopus/service/LiveImportClientImpl.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/scopus/service/LiveImportClientImpl.java:145]: Please add JavaDocs

File: dspace-api/src/test/data/dspaceFolder/assetstore/ads-ex.json
Comment (tdonohue) [inline dspace-api/src/test/data/dspaceFolder/assetstore/ads-ex.json:1]: Again, as mentioned in other PRs, these tests files should be moved to `dspace-server-webapp` alongside the ITs.

File: dspace-server-webapp/src/test/resources/test-config.properties
Comment (tdonohue) [inline dspace-server-webapp/src/test/resources/test-config.properties:23]: Again, if you move these files, these configs are no longer needed.

File: dspace/config/dspace.cfg
Comment (tdonohue) [inline dspace/config/dspace.cfg:1580]: These should all be moved to the new config you created in your other PR.  They also need comments to describe them.  In your comments, please make sure to describe what "ADS" means and what "EPO" means... abbreviations are not always easy to understand.

File: dspace/config/spring/api/ads-integration.xml
Comment (tdonohue) [inline dspace/config/spring/api/ads-integration.xml:83]: This field doesn't exist in DSpace.  Should it just be set to `dc.identifier.other`? or something similar?

File: dspace/config/spring/api/epo-integration.xml
Comment (tdonohue) [inline dspace/config/spring/api/epo-integration.xml:85]: This field doesn't exist.  Maybe just `dc.contributor`? or `dc.contributor.other`?

File: dspace/config/spring/api/epo-integration.xml
Comment (tdonohue) [inline dspace/config/spring/api/epo-integration.xml:94]: Again, field doesn't exist

**Review by tdonohue — CHANGES_REQUESTED** [review]
@Micheleboychuk : Gave this one a code review as well.  The comments here are very similar to in your other PRs. There's a lot of overlap (including some overlapping classes)...so we may want to try and move *one PR* forward first... and then rebase the others on that one.

**Comment by Micheleboychuk** [comment]
@tdonohue the next one to merge will be this one that I'm working on

File: dspace-api/src/main/java/org/dspace/importer/external/liveimportclient/service/LiveImportClient.java~main
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/liveimportclient/service/LiveImportClient.java~main:17]: This class strangely has an extension of `*.java-main`?  Maybe some sort of merge/rebase issue?

File: dspace-api/src/main/java/org/dspace/importer/external/metadatamapping/contributor/JsonPathMetadataProcessor.java~HEAD
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/metadatamapping/contributor/JsonPathMetadataProcessor.java~HEAD:19]: Another strange class with `*.java~HEAD` as the extension

File: dspace-api/src/main/java/org/dspace/importer/external/metadatamapping/contributor/JsonPathMetadataProcessor.java~main
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/metadatamapping/contributor/JsonPathMetadataProcessor.java~main:19]: And another file with an invalid extension.

File: dspace/config/modules/external-providers.cfg
Comment (tdonohue) [inline dspace/config/modules/external-providers.cfg:29]: Let's make sure to say what ADS is in this section header.  So, this is the NASA Astrophysics Data  System (ADS)

File: dspace/config/modules/external-providers.cfg
Comment (tdonohue) [inline dspace/config/modules/external-providers.cfg:38]: Same here, in the header we should spell out what EPO stands for.  Event just something as simple as saying: "European Patent Office (EPO)"

**Review by tdonohue — CHANGES_REQUESTED** [review]
@Micheleboychuk : I re-reviewed and tested this today.  First off, all my prior feedback seems to be addressed.  However, my testing wasn't successful & this PR includes some files with odd extensions that seem to not below.

(ADS) I found ADS is not working for me.  Any search returns an empty initial result.
no-results

In the logs on the backend, I see this exception:
```
2022-05-27 20:58:26,801 INFO  a14cfb2c-4e21-445c-bdb9-b1f53afe8843 a8eebd67-5fbf-4cec-bb12-a380b885348b org.dspace.app.rest.utils.DSpaceAPIRequestLoggingFilter @ Before request [GET /server/api/integration/externalsources/ads/entries] originated from /import-external?entity=none&sourceId=ads&query=place
2022-05-27 20:58:26,802 INFO  a14cfb2c-4e21-445c-bdb9-b1f53afe8843 a8eebd67-5fbf-4cec-bb12-a380b885348b org.dspace.importer.external.service.components.AbstractRemoteMetadataSource @ operation 7cc679de-42a5-419d-86c1-35119a2359a4 started
2022-05-27 20:58:27,274 ERROR a14cfb2c-4e21-445c-bdb9-b1f53afe8843 a8eebd67-5fbf-4cec-bb12-a380b885348b 
org.dspace.importer.external.liveimportclient.service.LiveImportClientImpl @ null
java.lang.RuntimeException: null
        at org.dspace.importer.external.liveimportclient.service.LiveImportClientImpl.executeHttpGetRequest(LiveImportClientImpl.java:77) [dspace-api-7.3-SNAPSHOT.jar:7.3-SNAPSHOT]
        at org.dspace.importer.external.ads.ADSImportMetadataSourceServiceImpl.search(ADSImportMetadataSourceServiceImpl.java:298) [dspace-api-7.3-SNAPSHOT.jar:7.3-SNAPSHOT]
        at org.dspace.importer.external.ads.ADSImportMetadataSourceServiceImpl$SearchByQueryCallable.call(ADSImportMetadataSourceServiceImpl.java:141) [dspace-api-7.3-SNAPSHOT.jar:7.3-SNAPSHOT]
        at org.dspace.importer.external.ads.ADSImportMetadataSourceServiceImpl$SearchByQueryCallable.call(ADSImportMetadataSourceServiceImpl.java:124) [dspace-api-7.3-SNAPSHOT.jar:7.3-SNAPSHOT]
        at org.dspace.importer.external.service.components.AbstractRemoteMetadataSource.retry(AbstractRemoteMetadataSource.java:171) [dspace-api-7.3-SNAPSHOT.jar:7.3-SNAPSHOT]
        at org.dspace.importer.external.ads.ADSImportMetadataSourceServiceImpl.getRecords(ADSImportMetadataSourceServiceImpl.java:81) [dspace-api-7.3-SNAPSHOT.jar:7.3-SNAPSHOT]
```

(EPO)  Currently, I'm not able to test EPO as I'm waiting on approval for a key.  However, I also noticed it's not testable as it require a `Patent` Entity (which doesn't exist in DSpace) along with several metadata fields which don't exist.   So, I think we need to figure out a better way to connect EPO into DSpace... maybe it can import to `Publication` Entities?  (I realize a Patent and a Publication are not entirely the same thing, but I'm not sure what else would work here... and I'm worried adding a new `Patent` type will be a lot more work.)

**Comment by Micheleboychuk** [comment]
hi @tdonohue , thanks for your feedback. Could you please restest again? Everything works for me now
nasa
epo
:

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/link/HalLinkFactory.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/link/HalLinkFactory.java:47]: I'm not entirely sure, but this small change _might_ be the cause to the templated links issue I described in my feedback.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@Micheleboychuk :  I'm still not able to get this PR to work.  However, the error has changed significantly.  It appears some change in this PR has _modified_ the format of HAL Links in a way that breaks all our search interfaces (both normal search & the entire MyDSpace page).  When I use `main`, everything works fine. But, if I install this PR on the backend, then the `/core/api/discover` endpoint starts returning templated links, see this screenshot:
templatedLinks

These `templated=true` links break all search functionality... so I cannot even load the MyDSpace page in order to test the rest of this PR.  I noted the area of code which might be causing this below.  I'm not 100% certain *why* this side effect is occurring, but templated links are NOT supported in DSpace at this time.

Another thing I realized in this code review is that we should investigate whether there's a way to *automatically disable* both EPO and ADS by default... as neither will work until the `key` configurations are filled out.  Perhaps we could use something like `@ConditionalOnProperty` for these services?  That way they are disabled until you add the necessary `key` configurations?

Finally, my feedback regarding EPO hasn't been addressed.  It's still configured to use an invalid Entity Type (Patent) and it still is configured to use several invalid metadata fields.  So, its configuration still needs to be fixed.

There are a few other minor notes inline below.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@Micheleboychuk : overall this is looking much better & I'm almost a +1.   I can confirm though that ADS works perfectly.  I've been unable to test EPO because the EPO team has not been able to provide me with an access key (still waiting).  That said, the EPO code looks good to me (and if you could test it on your end, I'd appreciate it).

However, I think we *must* have JavaDocs for the updates to HalLinkFactory (as the code is very confusing now). See notes inline below.

One other question (need not be solved in this PR):  Is there any way you know of that we can automatically turn *off* specific Import providers which are not configured yet?  I'm noticing that most of the new providers require some form of "access key".  

For example, it'd be nice if we had a way to auto-disable the ADS provider until `ads.key` is specified.  Same with EPO.  I'm not sure if there's any way to easily do this in the current code however (it'd be nice if we could use `@ConditionalOnProperty` for this, but I'm not sure if that works for these import sources yet).   Again, this isn't required in this PR, but I wanted to ask, as maybe we should create a separate ticket to handle this in the future (possibly in 7.4)

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/link/HalLinkFactory.java
Comment (abollini) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/link/HalLinkFactory.java:52]: I agree with @tdonohue having two methods that only differ for the parameters order is not acceptable so we should give a new name to one of the two,
The reason why we need two is actually because the existing one is buggy and lead in most case to a double encoding as reported here https://github.com/DSpace/DSpace/issues/8333

Unfortunately, the proper implementation that would rely completely on the "spring-hal-library" approach lead to the introduction of template link that is itself a good thing but right now would create trouble to angular as was previously noted by @tdonohue 

For this reason I suggest to add this alternative method as the preferred one and flag the other one as deprecated / add a javadoc referencing the reported bug.

In the context of the above bug fix we could get ride of the deprecated method but it is hard to deal with it in time for 7.3.

So the javadoc for the existing method would be
```
Please note that this method could lead to double encoding, you should use the new linkTo method instead. See also
```

the javadoc for the new method, to be renamed in getLinkTo could be
```
This is a shortcut to build a Link with the provided dummy method invocation result and the specified link relation.
```

Is this ok for you @tdonohue ? or should we just drop this new method and do what we need directly in the ExernalSourceEntryHalLinkFactory?

**Review by tdonohue — APPROVED** [review]
👍 Thanks @Micheleboychuk  (and @abollini ).  The updates look good to me, and I was able to successfully test both ADS and EPO (after getting access finally) today.  Both work perfectly.

I think this can be merged now (once all automated tests succeed obviously)

NOTE: I haven't received a response from either of you regarding whether you see any way to **auto-disable** specific providers (like EPO & ADS) if their access key is not configured.  Any thoughts on a way to achieve this?  Or should we just disable these providers all by default and document that they must be uncommented/enabled in order to use them?  -> We can move this discussion over to #8235 , but I'd still appreciate an answer as I want to ensure our documentation is clear on how to use these providers.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/link/HalLinkFactory.java
Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/link/HalLinkFactory.java]: Resolved, I see this was moved to `ExternalSourceEntryHalLinkFactory` for now, which is fine with me.

**Comment by tdonohue** [comment]
Merging with +1 approval. Thanks again @Micheleboychuk !

**Pull request merged**
